Replace StateDB

dylandreimerink opened the pull request:
This PR replaces the usage of StateDB with StateDB2 in L2 announcement and device controller / device manager.

```release-note
Replace StateDB with StateDB2
```

dylandreimerink committed:
l2announcement: Switch from StateDB to StateDB2

This commit switches the l2announcer and datapath/l2responder to use
StateDB2 instead of the original StateDB.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
devices_controller: Switched devices and routes from StateDB to StateDB2

This commit switches the devices controller over from the old StateDB
to the new StateDB2 implementation.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

giorio94 commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -83,136 +82,100 @@ func (p *l2ResponderReconciler) run(ctx context.Context) error {
 	ticker := time.NewTicker(5 * time.Minute)
 
+	tbl := p.params.L2AnnouncementTable
+	txn := p.params.StateDB.WriteTxn(tbl)
+	tracker, err := tbl.DeleteTracker(txn, "l2-responder-reconciler")
+	if err != nil {
+		txn.Abort()
+		return fmt.Errorf("delete tracker: %w", err)
+	}
+	txn.Commit()
```
Possible follow-up item. This logic looks common enough that could be abstracted in the statedb package.

joamaki committed:
statedb: Delete pkg/statedb and remove go-memdb dependency

pkg/statedb has been replaced by pkg/statedb2. Follow-up
PR will rename pkg/statedb2 to pkg/statedb.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -835,22 +835,18 @@ func (l2a *L2Announcer) processLeaderEvent(event leaderElectionEvent) error {
-	w := l2a.params.L2AnnounceTable.Writer(txn)
+	tbl := l2a.params.L2AnnounceTable
+	txn := l2a.params.StateDB.WriteTxn(tbl)
+	defer txn.Abort()
```
IIUC abort will cause a metric update even in case a commit already happened (while the same metric is not updated on `Commit()`.

joamaki replied:
Fixed this one!

giorio94 reviewed (approved):
I don't have much knowledge about the l2responder logic, but sig-agent changes look sensible to me.

joamaki committed:
statedb2: Fix metrics observing in Abort()

The metrics should be observed only if it was a write transaction
and it wasn't committed/aborted already.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
CODEOWNERS: statedb to statedb2

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb2: Add /statedb/dump REST API handler

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joestringer commented on pkg/l2announcer/l2announcer_test.go:
```diff
@@ -474,119 +467,17 @@ func TestPolicyRedundancy(t *testing.T) {
-
-// Test that when the last policy selecting a service is deleted, the selected service and proxy neighbor entries go
-// away.
-func TestPolicySoftDeleteService(t *testing.T) {
```
Do we need to revisit this test after this PR?

dylandreimerink replied:
We shouldn't have to. The old statedb had no way of tracking deletes so we had a custom soft delete mechanism which is just gone now, so we don't specifically have to test it anymore.

joestringer commented on pkg/l2announcer/l2announcer_test.go:
```diff
@@ -338,19 +335,16 @@ func TestHappyPathPermutations(t *testing.T) {
-					Origins:          []resource.Key{serviceKey(blueService())},
-					Deleted:          false,
-					Revision:         1,
+					L2AnnounceKey: tables.L2AnnounceKey{
+						IP:               netip.MustParseAddr(blueService().Spec.ExternalIPs[0]),
+						NetworkInterface: bluePolicy().Spec.Interfaces[0],
+					},
+					Origins: []resource.Key{serviceKey(blueService())},
```
We may want to have a bit of a discussion around this style of testing if changing the structure of data requires changing tests. I realize it's somewhat in the vein of "testing the translation from one table to another", but I think it's usually a weakness in the tests if changing the target code requires reworking the various structure declarations in the test code.

dylandreimerink replied:
I agree with the shortcomings of this method, but not really sure how I would change things. Do you suggest some sort of constructor or is there some other technique? Looking forward to that discussion.

joestringer reviewed (approved):
I briefly perused everything in my codeowners except `l2responder`. Approving for expedient removal of MPL-licensed code.
